spark/dl/src/main/scala/com/intel/analytics/bigdl/nn/Graph.scala

```diff
@@ -0,0 +1,283 @@
+        } else {
+          val nextGradOutputTable = n.element.gradInput.toTable
+          nextGradOutputTable[Tensor[T]](n.prevNodes.indexOf(curNode) + 1)
+        }
+
+        if (curGradOutput == null) {
+          curGradOutput = nextGradOutput
+        } else {
+          curGradOutput.add(nextGradOutput)
+        }
+      })
+      curNode.element.updateGradInput(inputsBP(i), curGradOutput)
+      i -= 1
+    }
+
+    gradInput = if (inputs.length == 1) {
+      inputs(0).element.gradInput
+    } else {
+      seqToTable(inputs.map(_.element.gradInput))
+    }
+
+    gradInput
+  }
+
+  override def accGradParameters(input: Activity, gradOutput: Activity, scale: Double): Unit = {
+    var j = outputs.length - 1
+    var i = executions.length - 1
+    if (outputs.length == 1) {
+      // Expect a tensor here
+      outputs(j).element.accGradParameters(inputsBP(i), gradOutput.toTensor)
+      j -= 1
+      i -= 1
+    } else {
+      val gradOutputTable = gradOutput.toTable
+      require(gradOutputTable.length() == outputs.length,
+        "gradient tensor number is not equal to outputs node number")
+      while(j >= 0) {
+        outputs(j).element.accGradParameters(inputsBP(i), gradOutputTable[Tensor[T]](j + 1))
+        j -= 1
+        i -= 1
+      }
+    }
+
+    while(i >= 0) {
+      val curNode = executions(i)
+      var curGradOutput : Tensor[T] = null
+      curNode.nextNodes.foreach(n => {
+        val nextGradOutput = if (n.prevNodes.length == 1) {
+          n.element.gradInput.toTensor
+        } else {
+          val nextGradOutputTable = n.element.gradInput.toTable
+          nextGradOutputTable[Tensor[T]](n.prevNodes.indexOf(curNode) + 1)
+        }
+
+        if (curGradOutput == null) {
+          curGradOutput = nextGradOutput
+        } else {
+          curGradOutput.add(nextGradOutput)
+        }
+      })
+      curNode.element.accGradParameters(inputsBP(i), curGradOutput)
+      i -= 1
+    }
+  }
+
+  // Add a dummy output node, to get an one end graph. So the nodes that are not dependent by
+  // the outputs will be excluded
+  private val dummyOutput = new ModuleNode[T](null)
+  outputs.foreach(_ -> dummyOutput)
+  private val backGraph = dummyOutput.graph(reverse = true)
+
+  // Build execution plan
+  private val executions = backGraph.topologySort.filter(_.element != null).reverse
+  modules.appendAll(executions.map(_.element.asInstanceOf[AbstractModule[Activity, Activity, T]]))
+
+  // Make input output nodes in executions be same order with how they are passed in
+  {
+    var i = 0
+    while(i < inputs.length) {
+      executions(i) = inputs(i)
+      i += 1
+    }
+
+    val offset = executions.length - outputs.length
+    i = 0
+    while(i < outputs.length) {
+      executions(i + offset) = outputs(i)
+      i += 1
+    }
+  }
+
+  private val inputsBP = new Array[Activity](executions.length)
+
+  // Check all inputs of the graph should be passed in
+  checkRoots
```

Comment:
Instead of checkRoots, we need to check the graph more thoroughly:
1) All outputs are connected to inputs, and all inputs are connected to outputs
2) All outputs have no nextNodes, and all inputs have no prevNodes
3) There are no cycles in the graph